vm/src/obj/objmemory.rs

```diff
@@ -1,74 +1,541 @@
-use super::objtype::{issubclass, PyClassRef};
+use std::fmt::Debug;
+
 use crate::bytesinner::try_as_bytes;
-use crate::common::hash::PyHash;
+use crate::obj::objbytes::{PyBytes, PyBytesRef};
+use crate::obj::objlist::{PyList, PyListRef};
+use crate::obj::{
+    objsequence::get_pos, objsequence::get_slice_range, objsequence::SequenceIndex,
+    objslice::PySliceRef, objstr::PyString, objtype::PyClassRef,
+};
 use crate::pyobject::{
-    ItemProtocol, PyClassImpl, PyContext, PyObjectRef, PyRef, PyResult, PyValue, TypeProtocol,
+    IdProtocol, PyClassImpl, PyContext, PyObjectRef, PyRef, PyResult, PyThreadingConstraint,
+    PyValue, TypeProtocol,
 };
-use crate::slots::Hashable;
-use crate::stdlib::array::PyArray;
+use crate::slots::{Bufferable, Hashable};
+use crate::stdlib::pystruct::_struct::FormatSpec;
 use crate::VirtualMachine;
+use crate::{common::hash::PyHash, pyobject::PyComparisonValue};
+use crossbeam_utils::atomic::AtomicCell;
+use itertools::Itertools;
+use num_bigint::BigInt;
+use num_traits::{One, Signed, ToPrimitive, Zero};
+use rustpython_common::borrow::{BorrowedValue, BorrowedValueMut};
+
+pub trait BufferProtocol: Debug + PyThreadingConstraint {
+    // TODO: return reference to avoid copy
+    fn get_options(&self) -> BufferOptions;
+    fn obj_bytes(&self) -> BorrowedValue<[u8]>;
+    fn obj_bytes_mut(&self) -> BorrowedValueMut<[u8]>;
+    fn release(&self);
+    fn is_resizable(&self) -> bool;
```

Comment:
I renamed as_bytes to obj_bytes because it should always return the full memory range for the original object. as_bytes may easier to be miss using.